Stabilize the integration-tests/qa-update

### <!-- Component -->

<!-- Zeebe- -->

### Description

tldr;
- the qa update tests try to pull images from DockerHub based on tags in Github for the newest version
- during the release of 8.8.3 there was an issue that a tag exists in github, but no image is yet published to dockerhub
- this is expected, but usually only for a short moment in time
- **we need to harden the module tests, e.g. rely on DockerHub or GitHub releases and not on git tags**
    - with https://github.com/camunda/camunda/pull/40801 the GitHub releases get created last

**Error observed:**
```
Caused by: com.github.dockerjava.api.exception.NotFoundException: Status 404: {"message":"manifest for camunda/zeebe:8.8.3 not found: manifest unknown: manifest **unknown"}**
```

**Image not listed on DockerHub**
```
% docker manifest inspect
camunda/zeebe:8.8.3 no such manifest: docker.io/camunda/zeebe:8.8.3
```

We use URL "https://api.github.com/repos/camunda/camunda/git/matching-refs/tags/8." to pull versions.

(see slack [conversation](https://camunda.slack.com/archives/C071KP5BTHB/p1762273810285779))

### Motivation

This issue highly effects everyone relying on the CI to merge anything. This tests are especially unstable during a release when new versions are published, but are not yet available on DockerHub. This might happen multiple time a month and potentially blocks the merge queue for a few hours. Recently, there has been an issue with the release process, causing this CI job to fail for ~1 day.

### Acceptance criteria

- [ ] Update Tests do not rely on github tags
- [ ] We reduce the likelihood for these tests to fail due to missing docker images


### Links

_No response_

### Breakdown

_No response_

--- a/zeebe/qa/update-tests/README.md
+++ b/zeebe/qa/update-tests/README.md
@@ -1,3 +1,5 @@
+see [Testing Guide](https://github.com/camunda/camunda/blob/main/docs/testing/acceptance.md#rolling-update-tests) for more information
+
 ## Pre-requisites
 
 To run the upgradability tests, a Docker image for Zeebe must be built with the tag 'current-test'. To do that you can run (in the camunda/zeebe dir):

--- a/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/RollingUpdateTest.java
+++ b/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/RollingUpdateTest.java
@@ -38,12 +38,14 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
 import org.agrona.CloseHelper;
 import org.awaitility.Awaitility;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.extension.RegisterExtension;
 import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.MethodSource;
 import org.testcontainers.DockerClientFactory;
 import org.testcontainers.containers.ContainerState;
@@ -57,6 +59,8 @@ import org.testcontainers.utility.DockerImageName;
  * versions.
  */
 final class RollingUpdateTest {
+
+  private static List<Arguments> cachedVersionMatrix;
 
   private static final BpmnModelInstance PROCESS =
       Bpmn.createExecutableProcess("process")
@@ -93,6 +97,13 @@ final class RollingUpdateTest {
 
   private final Collection<ZeebeVolume> volumes = new LinkedList<>();
 
+  static Stream<Arguments> versionMatrix() {
+    if (cachedVersionMatrix == null) {
+      cachedVersionMatrix = VersionCompatibilityMatrix.auto().toList();
+    }
+    return cachedVersionMatrix.stream();
+  }
+
   @BeforeEach
   public void setup() {
     cluster.getBrokers().values().forEach(this::configureBroker);
@@ -106,7 +117,7 @@ final class RollingUpdateTest {
   }
 
   @ParameterizedTest(name = "from {0} to {1}")
-  @MethodSource("io.camunda.zeebe.test.VersionCompatibilityMatrix#auto")
+  @MethodSource("versionMatrix")
   void shouldBeAbleToRestartContainerWithNewVersion(final String from, final String to) {
     // given
     updateAllBrokers(from);
@@ -137,7 +148,7 @@ final class RollingUpdateTest {
   }
 
   @ParameterizedTest(name = "from {0} to {1}")
-  @MethodSource("io.camunda.zeebe.test.VersionCompatibilityMatrix#auto")
+  @MethodSource("versionMatrix")
   void shouldReplicateSnapshotAcrossVersions(final String from, final String to) {
     // given
     updateAllBrokers(from);
@@ -204,7 +215,7 @@ final class RollingUpdateTest {
   }
 
   @ParameterizedTest(name = "from {0} to {1}")
-  @MethodSource("io.camunda.zeebe.test.VersionCompatibilityMatrix#auto")
+  @MethodSource("versionMatrix")
   void shouldPerformRollingUpdate(final String from, final String to) {
     // given
     updateAllBrokers(from);

--- a/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/VersionCompatibilityMatrix.java
+++ b/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/VersionCompatibilityMatrix.java
@@ -65,7 +65,7 @@ final class VersionCompatibilityMatrix {
    *       upgrade paths.
    * </ul>
    */
-  private static Stream<Arguments> auto() {
+  static Stream<Arguments> auto() {
     final var matrix = new VersionCompatibilityMatrix();
 
     if (System.getenv("ZEEBE_CI_CHECK_VERSION_COMPATIBILITY") != null) {

--- a/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/VersionCompatibilityMatrix.java
+++ b/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/VersionCompatibilityMatrix.java
@@ -42,6 +42,16 @@ final class VersionCompatibilityMatrix {
 
   private static final Logger LOG = LoggerFactory.getLogger(VersionCompatibilityMatrix.class);
 
+  private final VersionProvider versionProvider;
+
+  public VersionCompatibilityMatrix() {
+    versionProvider = new GithubVersionProvider();
+  }
+
+  public VersionCompatibilityMatrix(final VersionProvider versionProvider) {
+    this.versionProvider = versionProvider;
+  }
+
   /**
    * Automatically chooses the matrix to use depending on the environment where tests are run.
    *
@@ -56,22 +66,24 @@ final class VersionCompatibilityMatrix {
    * </ul>
    */
   private static Stream<Arguments> auto() {
+    final var matrix = new VersionCompatibilityMatrix();
+
     if (System.getenv("ZEEBE_CI_CHECK_VERSION_COMPATIBILITY") != null) {
-      return full();
+      return matrix.full();
     } else if (System.getenv("ZEEBE_CI_CHECK_CURRENT_VERSION_COMPATIBILITY") != null) {
-      return fromPreviousPatchesToCurrent();
+      return matrix.fromPreviousPatchesToCurrent();
     } else if (System.getenv("CI") != null) {
-      return fromFirstAndLastPatchToCurrent();
+      return matrix.fromFirstAndLastPatchToCurrent();
     } else {
-      return fromPreviousMinorToCurrent();
+      return matrix.fromPreviousMinorToCurrent();
     }
   }
 
-  private static Stream<Arguments> fromPreviousMinorToCurrent() {
+  public Stream<Arguments> fromPreviousMinorToCurrent() {
     return Stream.of(Arguments.of(VersionUtil.getPreviousVersion(), "CURRENT"));
   }
 
-  private static Stream<Arguments> fromPreviousPatchesToCurrent() {
+  public Stream<Arguments> fromPreviousPatchesToCurrent() {
     final var current = VersionUtil.getSemanticVersion().orElseThrow();
     return discoverVersions()
         .filter(version -> version.compareTo(current) < 0)
@@ -79,7 +91,7 @@ final class VersionCompatibilityMatrix {
         .map(version -> Arguments.of(version.toString(), "CURRENT"));
   }
 
-  private static Stream<Arguments> fromFirstAndLastPatchToCurrent() {
+  public Stream<Arguments> fromFirstAndLastPatchToCurrent() {
     final var current = VersionUtil.getSemanticVersion().orElseThrow();
 
     final var minAndMaxPatch =
@@ -92,7 +104,7 @@ final class VersionCompatibilityMatrix {
         Arguments.of(minAndMaxPatch.max().toString(), "CURRENT"));
   }
 
-  private static Stream<Arguments> full() {
+  public Stream<Arguments> full() {
     final var versions = discoverVersions().sorted().toList();
     final var latestVersionPerMinor = getLatestVersionPerMinor(versions);
     final var combinations =
@@ -128,16 +140,14 @@ final class VersionCompatibilityMatrix {
     return shard(combinations, index, total);
   }
 
-  private static Map<Integer, SemanticVersion> getLatestVersionPerMinor(
+  private Map<Integer, SemanticVersion> getLatestVersionPerMinor(
       final List<SemanticVersion> versions) {
-    final var latestVersionPerMinor =
-        versions.stream()
-            .collect(
-                Collectors.toMap(
-                    SemanticVersion::minor,
-                    Function.identity(),
-                    BinaryOperator.maxBy(Comparator.comparing(SemanticVersion::patch))));
-    return latestVersionPerMinor;
+    return versions.stream()
+        .collect(
+            Collectors.toMap(
+                SemanticVersion::minor,
+                Function.identity(),
+                BinaryOperator.maxBy(Comparator.comparing(SemanticVersion::patch))));
   }
 
   @VisibleForTesting
@@ -166,10 +176,10 @@ final class VersionCompatibilityMatrix {
    *     href="https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#list-matching-references--parameters">GitHub
    *     API</a>
    */
-  static Stream<SemanticVersion> discoverVersions() {
+  public Stream<SemanticVersion> discoverVersions() {
     final var versions =
-        fetchTags()
-            .map(Ref::toSemanticVersion)
+        versionProvider
+            .discoverVersions()
             .filter(version -> version != null && version.preRelease() == null)
             .toList();
 
@@ -178,7 +188,7 @@ final class VersionCompatibilityMatrix {
     // check for the latest versions of each minor.
     final var unreleasedVersions =
         getLatestVersionPerMinor(versions).values().stream()
-            .filter(version -> fetchRelease(version).isEmpty())
+            .filter(version -> !versionProvider.isReleased(version))
             .peek(
                 version ->
                     LOG.warn(
@@ -190,67 +200,85 @@ final class VersionCompatibilityMatrix {
     return versions.stream().filter(version -> !unreleasedVersions.contains(version));
   }
 
-  static Optional<Release> fetchRelease(final SemanticVersion tagName) {
-    final var endpoint =
-        URI.create(
-            "https://api.github.com/repos/camunda/camunda/releases/tags/" + tagName.toString());
-    try (final var httpClient = HttpClient.newHttpClient()) {
-      final var retry =
-          Retry.of(
-              "github-api",
-              RetryConfig.custom()
-                  .maxAttempts(10)
-                  .intervalFunction(IntervalFunction.ofExponentialBackoff())
-                  .build());
-      final var response =
-          retry.executeCallable(
-              () ->
-                  httpClient.send(
-                      HttpRequest.newBuilder().GET().uri(endpoint).build(),
-                      BodyHandlers.ofByteArray()));
-      return Optional.ofNullable(new ObjectMapper().readValue(response.body(), Release.class));
-    } catch (final Exception e) {
-      throw new RuntimeException(e);
+  static class GithubVersionProvider implements VersionProvider {
+
+    final Retry retry =
+        Retry.of(
+            "github-api",
+            RetryConfig.custom()
+                .maxAttempts(10)
+                .intervalFunction(IntervalFunction.ofExponentialBackoff())
+                .build());
+
+    @Override
+    public Stream<SemanticVersion> discoverVersions() {
+      return fetchTags().map(Ref::toSemanticVersion);
+    }
+
+    @Override
+    public boolean isReleased(final SemanticVersion version) {
+      return fetchRelease(version).isPresent();
+    }
+
+    private Optional<Release> fetchRelease(final SemanticVersion tagName) {
+      final var endpoint =
+          URI.create(
+              "https://api.github.com/repos/camunda/camunda/releases/tags/" + tagName.toString());
+      try (final var httpClient = HttpClient.newHttpClient()) {
+        final var response =
+            retry.executeCallable(
+                () ->
+                    httpClient.send(
+                        HttpRequest.newBuilder().GET().uri(endpoint).build(),
+                        BodyHandlers.ofByteArray()));
+
+        if (response.statusCode() == 404) {
+          return Optional.empty();
+        }
+
+        return Optional.ofNullable(new ObjectMapper().readValue(response.body(), Release.class));
+      } catch (final Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private Stream<Ref> fetchTags() {
+
+      final var endpoint =
+          URI.create("https://api.github.com/repos/camunda/camunda/git/matching-refs/tags/8.");
+      try (final var httpClient = HttpClient.newHttpClient()) {
+        final var response =
+            retry.executeCallable(
+                () ->
+                    httpClient.send(
+                        HttpRequest.newBuilder().GET().uri(endpoint).build(),
+                        BodyHandlers.ofByteArray()));
+
+        final var refs = new ObjectMapper().readValue(response.body(), Ref[].class);
+        return Stream.of(refs);
+      } catch (final Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    record Ref(String ref) {
+      SemanticVersion toSemanticVersion() {
+        return SemanticVersion.parse(ref.substring("refs/tags/".length())).orElse(null);
+      }
+    }
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    record Release(String tag_name, boolean prerelease) {
+      SemanticVersion toSemanticVersion() {
+        return SemanticVersion.parse(tag_name).orElse(null);
+      }
     }
   }
 
-  private static Stream<Ref> fetchTags() {
+  interface VersionProvider {
+    Stream<SemanticVersion> discoverVersions();
 
-    final var endpoint =
-        URI.create("https://api.github.com/repos/camunda/camunda/git/matching-refs/tags/8.");
-    try (final var httpClient = HttpClient.newHttpClient()) {
-      final var retry =
-          Retry.of(
-              "github-api",
-              RetryConfig.custom()
-                  .maxAttempts(10)
-                  .intervalFunction(IntervalFunction.ofExponentialBackoff())
-                  .build());
-      final var response =
-          retry.executeCallable(
-              () ->
-                  httpClient.send(
-                      HttpRequest.newBuilder().GET().uri(endpoint).build(),
-                      BodyHandlers.ofByteArray()));
-
-      final var refs = new ObjectMapper().readValue(response.body(), Ref[].class);
-      return Stream.of(refs);
-    } catch (final Exception e) {
-      throw new RuntimeException(e);
-    }
-  }
-
-  @JsonIgnoreProperties(ignoreUnknown = true)
-  record Ref(String ref) {
-    SemanticVersion toSemanticVersion() {
-      return SemanticVersion.parse(ref.substring("refs/tags/".length())).orElse(null);
-    }
-  }
-
-  @JsonIgnoreProperties(ignoreUnknown = true)
-  private record Release(String tag_name, boolean prerelease) {
-    SemanticVersion toSemanticVersion() {
-      return SemanticVersion.parse(tag_name).orElse(null);
-    }
+    boolean isReleased(SemanticVersion version);
   }
 }

--- a/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/VersionCompatibilityMatrix.java
+++ b/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/VersionCompatibilityMatrix.java
@@ -21,6 +21,8 @@ import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
 import java.net.http.HttpResponse.BodyHandlers;
 import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
 import java.util.Optional;
 import java.util.SequencedCollection;
 import java.util.function.BinaryOperator;
@@ -28,6 +30,8 @@ import java.util.function.Function;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import org.junit.jupiter.params.provider.Arguments;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Provides combinations of versions that should be compatible with each other. This matrix is used
@@ -35,6 +39,8 @@ import org.junit.jupiter.params.provider.Arguments;
  */
 @SuppressWarnings("unused")
 final class VersionCompatibilityMatrix {
+
+  private static final Logger LOG = LoggerFactory.getLogger(VersionCompatibilityMatrix.class);
 
   /**
    * Automatically chooses the matrix to use depending on the environment where tests are run.
@@ -88,13 +94,7 @@ final class VersionCompatibilityMatrix {
 
   private static Stream<Arguments> full() {
     final var versions = discoverVersions().sorted().toList();
-    final var latestVersionPerMinor =
-        versions.stream()
-            .collect(
-                Collectors.toMap(
-                    SemanticVersion::minor,
-                    Function.identity(),
-                    BinaryOperator.maxBy(Comparator.comparing(SemanticVersion::patch))));
+    final var latestVersionPerMinor = getLatestVersionPerMinor(versions);
     final var combinations =
         versions.stream()
             .filter(version -> version.minor() > 0)
@@ -128,6 +128,18 @@ final class VersionCompatibilityMatrix {
     return shard(combinations, index, total);
   }
 
+  private static Map<Integer, SemanticVersion> getLatestVersionPerMinor(
+      final List<SemanticVersion> versions) {
+    final var latestVersionPerMinor =
+        versions.stream()
+            .collect(
+                Collectors.toMap(
+                    SemanticVersion::minor,
+                    Function.identity(),
+                    BinaryOperator.maxBy(Comparator.comparing(SemanticVersion::patch))));
+    return latestVersionPerMinor;
+  }
+
   @VisibleForTesting
   static <T> Stream<T> shard(final SequencedCollection<T> list, final int index, final int total) {
     if (list.size() < total) {
@@ -146,17 +158,64 @@ final class VersionCompatibilityMatrix {
    * Discovers Zeebe versions that aren't pre-releases. Sourced from the GitHub API and can fail on
    * network issues. Includes all versions since 8.0.
    *
+   * <p>For the latest patch version of each minor version, verifies that the version has been
+   * published as a GitHub release before including it. This ensures that only versions with
+   * published artifacts are included in the compatibility matrix.
+   *
    * @see <a
    *     href="https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#list-matching-references--parameters">GitHub
    *     API</a>
    */
   static Stream<SemanticVersion> discoverVersions() {
-    @JsonIgnoreProperties(ignoreUnknown = true)
-    record Ref(String ref) {
-      SemanticVersion toSemanticVersion() {
-        return SemanticVersion.parse(ref.substring("refs/tags/".length())).orElse(null);
-      }
+    final var versions =
+        fetchTags()
+            .map(Ref::toSemanticVersion)
+            .filter(version -> version != null && version.preRelease() == null)
+            .toList();
+
+    // Filter out unreleased versions which might not have the right artifacts like maven/docker
+    // published yet. The GitHub API for fetching releases is inflexible, thus we optimize to only
+    // check for the latest versions of each minor.
+    final var unreleasedVersions =
+        getLatestVersionPerMinor(versions).values().stream()
+            .filter(version -> fetchRelease(version).isEmpty())
+            .peek(
+                version ->
+                    LOG.warn(
+                        "Latest patch {} for minor version 8.{} has no corresponding GitHub release yet. Excluding from compatibility matrix.",
+                        version,
+                        version.minor()))
+            .collect(Collectors.toSet());
+
+    return versions.stream().filter(version -> !unreleasedVersions.contains(version));
+  }
+
+  static Optional<Release> fetchRelease(final SemanticVersion tagName) {
+    final var endpoint =
+        URI.create(
+            "https://api.github.com/repos/camunda/camunda/releases/tags/" + tagName.toString());
+    try (final var httpClient = HttpClient.newHttpClient()) {
+      final var retry =
+          Retry.of(
+              "github-api",
+              RetryConfig.custom()
+                  .maxAttempts(10)
+                  .intervalFunction(IntervalFunction.ofExponentialBackoff())
+                  .build());
+      final var response =
+          retry.executeCallable(
+              () ->
+                  httpClient.send(
+                      HttpRequest.newBuilder().GET().uri(endpoint).build(),
+                      BodyHandlers.ofByteArray()));
+      return Optional.ofNullable(new ObjectMapper().readValue(response.body(), Release.class));
+    } catch (final Exception e) {
+      throw new RuntimeException(e);
     }
+  }
+
+  private static Stream<Ref> fetchTags() {
+
     final var endpoint =
         URI.create("https://api.github.com/repos/camunda/camunda/git/matching-refs/tags/8.");
     try (final var httpClient = HttpClient.newHttpClient()) {
@@ -173,12 +232,25 @@ final class VersionCompatibilityMatrix {
                   httpClient.send(
                       HttpRequest.newBuilder().GET().uri(endpoint).build(),
                       BodyHandlers.ofByteArray()));
+
       final var refs = new ObjectMapper().readValue(response.body(), Ref[].class);
-      return Stream.of(refs)
-          .map(Ref::toSemanticVersion)
-          .filter(version -> version != null && version.preRelease() == null);
+      return Stream.of(refs);
     } catch (final Exception e) {
       throw new RuntimeException(e);
+    }
+  }
+
+  @JsonIgnoreProperties(ignoreUnknown = true)
+  record Ref(String ref) {
+    SemanticVersion toSemanticVersion() {
+      return SemanticVersion.parse(ref.substring("refs/tags/".length())).orElse(null);
+    }
+  }
+
+  @JsonIgnoreProperties(ignoreUnknown = true)
+  private record Release(String tag_name, boolean prerelease) {
+    SemanticVersion toSemanticVersion() {
+      return SemanticVersion.parse(tag_name).orElse(null);
     }
   }
 }

--- a/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/VersionCompatibilityMatrixTest.java
+++ b/zeebe/qa/update-tests/src/test/java/io/camunda/zeebe/test/VersionCompatibilityMatrixTest.java
@@ -10,14 +10,49 @@ package io.camunda.zeebe.test;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.params.provider.Arguments.arguments;
 
+import io.camunda.zeebe.test.VersionCompatibilityMatrix.VersionProvider;
+import io.camunda.zeebe.util.SemanticVersion;
 import java.util.LinkedList;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
+import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.MethodSource;
 
 class VersionCompatibilityMatrixTest {
+
+  @Test
+  void filterPreReleaseVersion() {
+    final VersionCompatibilityMatrix matrix =
+        new VersionCompatibilityMatrix(
+            new StaticVersionProvider(
+                Set.of("8.7.11", "8.8.2", "8.8.2-optimize"),
+                Set.of("8.7.11", "8.8.2", "8.8.2-optimize")));
+
+    final var versions =
+        matrix.discoverVersions().map(SemanticVersion::toString).collect(Collectors.toSet());
+
+    assertThat(versions).isEqualTo(Set.of("8.7.11", "8.8.2"));
+  }
+
+  @Test
+  void filterUnreleasedLatestVersion() {
+    final VersionCompatibilityMatrix matrix =
+        new VersionCompatibilityMatrix(
+            new StaticVersionProvider(
+                Set.of("8.7.11", "8.8.0", "8.8.1", "8.8.2", "8.8.2-optimize"),
+                Set.of("8.7.11", "8.8.0", "8.8.1")));
+
+    final var versions =
+        matrix.discoverVersions().map(SemanticVersion::toString).collect(Collectors.toSet());
+
+    assertThat(versions).isEqualTo(Set.of("8.7.11", "8.8.0", "8.8.1"));
+  }
+
   @ParameterizedTest(name = "Sharding {0} elements into {1} shards")
   @MethodSource("sizeAndShards")
   void shouldShardCompletely(final int size, final int totalShards) {
@@ -40,5 +75,34 @@ class VersionCompatibilityMatrixTest {
                     .boxed()
                     .filter(shards -> shards < size)
                     .map(shards -> arguments(size, shards)));
+  }
+
+  static class StaticVersionProvider implements VersionProvider {
+
+    private final Set<SemanticVersion> versions;
+    private final Set<SemanticVersion> releasedVersions;
+
+    public StaticVersionProvider(final Set<String> versions, final Set<String> releasedVersions) {
+      this.versions =
+          versions.stream()
+              .map(SemanticVersion::parse)
+              .flatMap(Optional::stream)
+              .collect(Collectors.toSet());
+      this.releasedVersions =
+          releasedVersions.stream()
+              .map(SemanticVersion::parse)
+              .flatMap(Optional::stream)
+              .collect(Collectors.toSet());
+    }
+
+    @Override
+    public Stream<SemanticVersion> discoverVersions() {
+      return versions.stream();
+    }
+
+    @Override
+    public boolean isReleased(final SemanticVersion version) {
+      return releasedVersions.contains(version);
+    }
   }
 }

